doc/decisions/binary.md

```diff
@@ -0,0 +1,40 @@
+# Binary
+
+## Problem
+
+binary
+
+- is the only metadata that allows more values (instead of making the validation stricter)
+- is the only API that modifies metadata
+
+## Constraints
+
+- simplify API
+
+## Assumptions
+
+byte-array values are a rarely used feature in configuration settings
+
+## Considered Alternatives
+
+- flag for indicating if a value is binary
+
+## Decision
+
+- make keys with untyped value per default (no assumption about type, i.e., byte-array)
+- all current types are not binary, so type = string is the way to indicate a string doesn't
+  - contain null bytes
+  - is not null (not only indicator of structure without value)
```

Comment:
Where do we need this requirement? Is this a CORBA string defined like this?

Reply:
> - Characters, as defined in ISO Latin-1 (8859.1) and other single- or multi-byte character sets
> - [...]
> - A string type, which consists of a variable-length array of characters; the length of the string is a non-negative integer, and is available at run-time. The length may have a maximum bound defined.
>
> ~ CORBA Specification Version 3.4, Part 1, Section 5.2.4.1 Basic Types

So yes, CORBA (kind of) defines this. However, I really wouldn't use this definition directly. The definition for "string" is fine, but relies on the definition of a "character", which is incredibly vague, and almost useless. The way I read this, it's not even necessary that all characters use the same encoding.

The actual reason, why I suggested this addition is that without an underlying text encoding a string is hardly a string at all. The encoding doesn't have to be stored, it could be contextual knowledge the application has, be derived from the metadata, etc. The encoding could even be different for each key. The only thing that matters IMO is that on somewhere, someone agreed on a way of turning the byte sequence into text. Otherwise, whats the point...

> Where do we need this requirement?

For example in `kdb get /some/key`. We print the key value and assume the terminal is configured with the correct text encoding. We don't care what this encoding actually is, but it has to exist, otherwise we will always print garbage.